Implement auto-detect, completion filtering, and visual distinction

allisoneer opened the pull request:
## What problem(s) was I solving?

**User experience problems:**
- Users had to manually configure `GWT_GIT_DIR` or set a default repository even when working within a git repository, causing friction when switching between repos or working in subdirectories
- Shell completion didn't distinguish between branch types (worktrees, local branches, remote branches), making it hard to see which branches were already checked out
- `gwt rm` completion showed all branches including main, risking accidental removal attempts
- Error messages were generic and didn't help users understand what went wrong

**Code quality and maintainability problems:**
- Monolithic 1,449-line `gwt.py` was difficult to maintain, navigate, and test in isolation
- CI workflow only ran flake8 on `gwt.py`, missing lint/format/type issues in other files
- No testing matrix across Python versions
- Build artifacts weren't validated before declaring CI success

## What user-facing changes did I ship?

1. **Auto-detection of git repository**: `gwt` now automatically detects the git repository from the current working directory, eliminating the need to set `GWT_GIT_DIR` when working within a repo. Works from repo root, subdirectories, and existing worktrees. Priority order: CLI argument → auto-detect → `GWT_GIT_DIR` → config file.

2. **Visual distinction in shell completion**: Branch names in shell completion now show their type:
   - **Bash**: Symbol prefixes (● worktree, ○ local, ⊙ remote) that are automatically stripped when selected
   - **Fish**: Tab-separated descriptions showing branch type using native completion support

3. **Smarter completion filtering**: 
   - `gwt switch` shows all branches (including main)
   - `gwt rm` only shows removable worktrees (excluding main)
   - Prevents accidental attempts to remove main worktree

4. **Better error messages**: Source-specific error messages (E001, E002, E003) with helpful hints:
   - E001: No repo detected and no valid fallbacks
   - E002: `GWT_GIT_DIR` points to invalid path
   - E003: Config `default_repo` is invalid

5. **No breaking changes**: Complete backward compatibility maintained through careful API re-exports. All 23 tests pass, including new validation tests for module imports and public API surface.

6. **Improved code quality**: Comprehensive CI now enforces formatting, linting, type checking, and tests across Python 3.10-3.13, ensuring higher quality releases.

## How I implemented it

### Auto-detection (gwtlib/resolution.py)
- Added `auto_detect_git_dir()` function using `git rev-parse --git-common-dir` for reliable detection across repo root, subdirectories, worktrees, and submodules
- Refactored `get_git_dir()` into `get_git_dir_with_source()` that returns `(git_dir, source, meta)` tuple for detailed error handling
- Added `_normalize_repo_path()` helper to auto-append `.git` for non-bare repositories
- Updated main() to use new detection system with source-aware error messages

### Completion filtering (gwtlib/branches.py, gwt.sh:63-94, gwt.fish:82-91)
- Modified `list_all_branches()` to accept `mode` parameter: "all", "local", "worktrees"
- For mode="worktrees", exclude main worktree and only return branches with existing worktrees
- Split bash/fish completion routing: `switch` uses `--branches all`, `remove` uses `--branches worktrees`
- Added separate completion paths in shell wrappers for different subcommands

### Visual distinction (gwtlib/display.py, gwt.sh:98-112)
- Added `--annotate` argument to `list` command with choices: "none", "bash", "fish"
- Implemented `print_branch()` helper function in `list_all_branches()` to format output based on annotation type
- Bash: Prefix branches with symbols (● ○ ⊙) that are stripped by `_gwt_strip_visual()` wrapper function before passing to gwt.py
- Fish: Output tab-separated format (branch\tdescription) using native completion support
- Both formats maintain branch type ordering: worktrees → local → remote

### Modular refactoring (gwt.py → gwtlib/ package)
Transformed monolithic 1,449-line `gwt.py` into clean modular architecture:

- **gwt.py** (45 lines): Thin shim preserving PEP 723 header, re-exports public API for backward compatibility
- **gwtlib/** package (11 modules, 1,438 total lines):
  * `config.py` - Configuration management and HAS_TOML centralization
  * `git_ops.py` - Standardized git command wrappers
  * `paths.py` - Path utilities for worktrees and display
  * `branches.py` - Branch detection and remote operations
  * `parsing.py` - Git worktree list parsing and reconciliation
  * `display.py` - Formatting and listing functions
  * `worktrees.py` - Worktree creation, switching, and removal
  * `resolution.py` - Git directory auto-detection
  * `cli.py` - CLI argument parsing and dispatch
  * `api.py` - Public API aggregator for test compatibility
  * `__init__.py` - Package initialization

Each module stays well under 800-line target (largest: 334 lines). All git subprocess calls standardized through `git_ops` wrappers.

### CI modernization (.github/workflows/ci.yml, justfile)
Replaced limited flake8-only CI with comprehensive 4-job workflow delegating to justfile recipes:

**Justfile changes:**
- Added CI detection (`CI=true`, `OUTPUT_MODE`, `shellcheck_format`) for environment-aware execution
- Expanded fmt/lint/type-check scope from `gwt.py` to all Python sources (`gwt.py gwtlib tests`)
- Added `check-shell` recipe for comprehensive shellcheck coverage of all shell scripts
- Added `build-validate` recipe with smoke test (`./dist/gwt --help`)
- Fixed `build` recipe to copy `gwtlib/` directory alongside `gwt.py`

**CI workflow:**
1. **Lint job**: Format check, lint (ruff), type check (ty, non-blocking) on Python 3.12
2. **Test matrix**: pytest across Python 3.10, 3.11, 3.12, 3.13 (fail-fast: false)
3. **ShellCheck job**: Install ShellCheck 0.11.0, run `just check-shell` on all bash/sh scripts
4. **Build job**: Validate build artifacts with smoke test (depends on lint, test, shellcheck)

All checks now run via justfile for local/CI parity. Bumped minimum Python support from 3.9+ to 3.10+ to align with modern tooling.

### Import fixes (all gwtlib modules, tests/)
Applied ruff import sorting (I001) and removed unused imports (F401) across all modules to comply with expanded lint scope. Ensures consistent import ordering and eliminates dead code.

### Test coverage
- **Original feature tests** (tests/test_cli.py:48-423, tests/test_worktree_management.py:117-195): 11 tests covering auto-detection, priority ordering, error handling, annotation formats, worktree filtering
- **Refactoring validation tests** (tests/test_gwtlib_imports.py, tests/test_public_api_re_exports.py): 2 tests ensuring all gwtlib modules import cleanly and public API is properly re-exported from gwt.py
- **Existing tests updated**: Absolute paths and proper environment isolation to avoid auto-detection interference
- **Total coverage**: 23 tests, all passing

## How to verify it

- [x] I have ensured `just check` passes (formatting, linting, type checking)
- [x] I have ensured `just test` passes (23 tests, all passing)
- [x] Manual verification: Test auto-detection by running `gwt list` from within a git repo (without `GWT_GIT_DIR` set)
- [x] Manual verification: Test bash completion with `gwt switch <TAB>` to see visual symbols
- [x] Manual verification: Test fish completion with `gwt switch <TAB>` to see descriptions  
- [x] Manual verification: Test `gwt rm <TAB>` confirms main is excluded from suggestions
- [x] Manual verification: Verify backward compatibility by running `gwt` in existing workflows

## Description for the changelog

Added auto-detection of git repository from current directory, visual distinction for branch types in shell completion (worktrees, local, remote), and filtered completion for `gwt rm` to exclude main worktree. Improved error messages with source-specific hints. Refactored monolithic `gwt.py` into modular `gwtlib` package (11 focused modules) for better maintainability. Modernized CI with comprehensive 4-job workflow (lint, test matrix across Python 3.10-3.13, shellcheck, build validation) delegating to justfile for local/CI parity. All changes maintain 100% backward compatibility.

Commit:
Implement auto-detect, completion filtering, and visual distinction

Add three major improvements to gwt:

1. Auto-detect git repository from current directory
   - Uses git rev-parse --git-common-dir for reliable detection
   - Reversed priority: auto-detect → env var → config
   - Works from repo root, subdirectories, and worktrees
   - Source-specific error messages (E001, E002, E003) with helpful hints

2. Filter gwt rm completion to show only removable branches
   - Excludes main worktree from rm suggestions
   - Split bash/fish completion routing for switch vs rm
   - Empty completion when no worktrees exist

3. Add visual distinction for branch types in shell completion
   - Bash: Symbol prefixes (● worktree, ○ local, ⊙ remote) with runtime stripping
   - Fish: Tab-separated descriptions using native support
   - New --annotate flag for list command

All changes include comprehensive test coverage (11 new tests).
All tests pass (21 total). Resolves ENG-233.

Commit:
Refactor gwt.py into modular gwtlib package structure

Transform monolithic 1,449-line gwt.py into clean modular architecture:

- Reduce gwt.py to 45-line thin shim that preserves PEP 723 header
  and re-exports public API for backward compatibility
- Create gwtlib/ package with 11 focused modules organized by concern:
  * config.py - Configuration and HAS_TOML centralization
  * git_ops.py - Standardized git command wrappers
  * paths.py - Path utilities for worktrees and display
  * branches.py - Branch detection and remote operations
  * parsing.py - Git worktree list parsing and reconciliation
  * display.py - Formatting and listing functions
  * worktrees.py - Worktree creation, switching, and removal
  * resolution.py - Git directory auto-detection
  * cli.py - CLI argument parsing and dispatch
  * api.py - Public API aggregator for test compatibility
- Add validation tests for module imports and API re-exports
- Standardize all git subprocess calls through git_ops wrappers
- Maintain 100% backward compatibility (all 23 tests pass)

All modules stay well under 800-line target (largest: 334 lines).
Preserves original behavior while improving maintainability.

Commit:
Fix import sorting and remove unused imports

Apply ruff import sorting (I001) and remove unused imports (F401)
across gwtlib modules and tests to comply with expanded lint scope.

Commit:
Modernize CI and align with justfile

Replace flake8-only CI with comprehensive 4-job workflow that delegates
to justfile recipes. Expand all checks to cover gwt.py, gwtlib/, and
tests/ instead of just gwt.py.

Changes:
- Add CI detection (CI=true, OUTPUT_MODE, shellcheck_format) to justfile
- Expand fmt/lint/type-check scope from gwt.py to all Python sources
- Add check-shell recipe for comprehensive shellcheck coverage
- Add build-validate recipe with smoke test
- Fix build recipe to copy gwtlib directory
- Replace CI workflow with 4 jobs: lint, test (matrix 3.10-3.13),
  shellcheck, and build validation
- Bump Python support from 3.9+ to 3.10+

All checks now run via justfile for local/CI parity. Local verification
shows all checks passing: lint, format, type-check, tests, shellcheck,
and build validation.

ericaltendorf commented on justfile:
```diff
@@ -1,9 +1,17 @@
 set shell := ["bash", "-euo", "pipefail", "-c"]
 
+# CI/output mode detection (inspired by monorepo pattern)
+ci := env("CI", "false")
```
this is an actual environment variable, right?  Maybe use a prefix on the env var name to reduce risk of collisions in the global namespace?

allisoneer replied:
This doesn't set it, it just reads it. "false" is default if it's not already set. If anything else (in global environment variable) sets CI to true, then we read it as true and don't use the default.

Commit:
Migrate installation to AppDir layout with XDG compliance

Move Python sources (gwt.py + gwtlib/) from ~/.local/bin to AppDir at
~/.local/share/gwt (respecting XDG_DATA_HOME), keeping only thin shell
wrappers in ~/.local/bin. This fixes ModuleNotFoundError after the
modular refactoring by preserving co-location of gwt.py and gwtlib/.

Changes:
- install.sh: Install to AppDir, clean up old gwt.py, enhance uv messaging
- gwt.sh: Add AppDir resolution, uv/python3 fallback, zsh bashcompinit
- gwt.fish: Rewrite with AppDir helpers and uv/python3 fallback
- README.md: Document AppDir layout, manual install, and uninstall steps

Shell integration (cd/export) and tab completion preserved. Automatic
cleanup of legacy installations from ~/.local/bin.

ericaltendorf commented:
I understand, it's a read dependency not a write dependency, but it's still
a dependency of information being communicated through a channel which
could easily conflict with some other channel since they're in a global
namespace.

Not super important, just seems like a potential gotcha, but mostly I am
very suspicious of env vars :)

On Tue, Nov 4, 2025 at 9:14 AM Allison Durham ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In justfile
> <https://github.com/ericaltendorf/gwt/pull/4#discussion_r2491380895>:
>
> > @@ -1,9 +1,17 @@
>  # Use bash with strict flags
>  set shell := ["bash", "-euo", "pipefail", "-c"]
>
> +# CI/output mode detection (inspired by monorepo pattern)
> +ci := env("CI", "false")
>
> This doesn't set it, it just reads it. "false" is default if it's not
> visible. If anything else (in global environment variable) sets CI to true,
> then we read it as true and don't use the default.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ericaltendorf/gwt/pull/4#discussion_r2491380895>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARPZIFRX6BDMMUXZBMF6CJD33DNF5AVCNFSM6AAAAACKZ526POVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTIMJXG4YDONBTGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
